In [1]:
import re
import numpy as np
import pandas as pd
df = pd.read_csv('2014_a_2024-1.csv')

/tmp/ipykernel_41736/4015083117.py:4: DtypeWarning: Columns (1,2,3,4,5,6,7,8,9,11,13,15,17,18,19,20,21,22,23,24,25,26,27,28,30,32,33,34,36,38,39,40,41,42,43,44,45,46,47,48,49,50,51,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('2014_a_2024-1.csv')


In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1128678 entries, 0 to 1128677
Data columns (total 77 columns):
 #   Column                                            Non-Null Count    Dtype  
---  ------                                            --------------    -----  
 0   Data de cadastro                                  1128678 non-null  object 
 1   Canal de atendimento                              592610 non-null   object 
 2   Denúncia emergencial                              461710 non-null   object 
 3   Denunciante                                       946341 non-null   object 
 4   Cenário da violação                               590210 non-null   object 
 5   País                                              403786 non-null   object 
 6   UF                                                889787 non-null   object 
 7   Município                                         871949 non-null   object 
 8   Frequência                                        784879 non-null   obje

In [23]:
print(df["Violações"])
contagem = df["Violações"].value_counts()
print(contagem)

0                INTEGRIDADE>PSIQUICA>AMEACA OU COACAO
1                    INTEGRIDADE>FISICA>TORTURA FISICA
2                       INTEGRIDADE>PSIQUICA>EXPOSICAO
3          INTEGRIDADE>FISICA>AGRESSAO OU VIAS DE FATO
4                       INTEGRIDADE>PSIQUICA>DIFAMACAO
                              ...                     
1128673          INTEGRIDADE>PSIQUICA>TORTURA PSIQUICA
1128674                        INTEGRIDADE>NEGLIGENCIA
1128675    INTEGRIDADE>FISICA>AGRESSAO OU VIAS DE FATO
1128676         LIBERDADE>SEXUAL>ESTUPRO DE VULNERAVEL
1128677       LIBERDADE>SEXUAL>PSIQUICA>ASSEDIO SEXUAL
Name: Violações, Length: 1128678, dtype: object
Violações
INTEGRIDADE>PSIQUICA>AMEACA OU COACAO                                                                                                                         55538
INTEGRIDADE>PSIQUICA>TORTURA PSIQUICA                                                                                                                         52095
INTEGRIDA

In [4]:
# Contabilizando diretamente as linhas que contêm a palavra "ESTUPRO"
contagem_estupro = df['Violações'].str.contains("ESTUPRO", na=False).sum()

print(f"\nTotal de linhas que contêm a palavra 'ESTUPRO': {contagem_estupro}")


Total de linhas que contêm a palavra 'ESTUPRO': 54359


In [5]:
contagem_trafico_mulher = df['Violações'].str.contains("TRAFICO DE MULHERES", na=False).sum()
print(f"\nTotal de linhas que contêm a palavra 'TRAFICO DE MULHERES': {contagem_trafico_mulher}")


Total de linhas que contêm a palavra 'TRAFICO DE MULHERES': 100


In [6]:
# Função para pegar a primeira palavra antes de ">", "." ou ";"
def get_first_word(text):
    if isinstance(text, str):
        return re.split(r'[>.;]', text)[0]
    return None

df['primeira_palavra'] = df['Violações'].apply(get_first_word)

primeiras_palavras_distintas = df['primeira_palavra'].dropna().value_counts()
print(primeiras_palavras_distintas)

primeira_palavra
INTEGRIDADE                                 353260
DANO EMOCIONAL/DIMINUICAO DA AUTO-ESTIMA     97143
VIOLENCIA DOMESTICA E FAMILIAR               89734
AMEACA                                       88992
LESAO CORPORAL LEVE                          79823
INJURIA                                      45334
VIOLENCIA_FISICA                             40643
DIFAMACAO                                    34857
VIOLENCIA PATRIMONIAL                        29599
CALUNIA                                      27684
LIBERDADE                                    22630
ESTUPRO                                      20485
PERSEGUICOES                                 17368
VIOLENCIA_PSICOLOGICA                        16380
LESAO CORPORAL GRAVE                         12304
CARCERE PRIVADO                              10519
OUTRO                                         9360
VIOLENCIA MORAL                               8272
FEMINICIDIO                                   6368
VIOLENCIA INST

In [18]:
# Função para pegar o texto após a primeira palavra
def get_following_text(text):
    if isinstance(text, str) and text.startswith("LIBERDADE"):
        return re.split(r'[>.;]', text, 1)[1] if len(re.split(r'[>.;]', text)) > 1 else None
    return None

# Aplicando a função para pegar os valores após "INTEGRIDADE"
df['seguinte_integridade'] = df['Violações'].apply(get_following_text)

valores_seguinte_integridade = df['seguinte_integridade'].dropna().value_counts()
print(valores_seguinte_integridade)

seguinte_integridade
DIREITOS INDIVIDUAIS>AUTONOMIA DE VONTADE                                                                      4057
SEXUAL>FISICA>ESTUPRO                                                                                          3549
SEXUAL>PSIQUICA>ASSEDIO SEXUAL                                                                                 2415
DIREITOS INDIVIDUAIS>CARCERE PRIVADO                                                                           2302
DIREITOS INDIVIDUAIS>LIBERDADE DE IR VIR PERMANECER                                                            1417
SEXUAL>FISICA>ABUSO / IMPORTUNACAO SEXUAL FISICA                                                               1398
SEXUAL>PSIQUICA>ABUSO / IMPORTUNACAO SEXUAL PSIQUICA                                                           1274
SEXUAL>PSIQUICA>ABUSO SEXUAL PSIQUICO                                                                          1255
SEXUAL>FISICA>ABUSO SEXUAL FISICO                  

In [19]:
# Função para pegar o texto após a primeira palavra
def get_following_text(text):
    if isinstance(text, str) and text.startswith("LIBERDADE"):
        return re.split(r'[>.;]', text, 1)[1] if len(re.split(r'[>.;]', text)) > 1 else None
    return None

# Aplicando a função para pegar os valores após "INTEGRIDADE"
df['seguinte_integridade'] = df['Violações'].apply(get_following_text)
# Formatar cada linha para começar com "INTEGRIDADE>"
formatted_output = "\n".join([f"LIBERDADE>{index} {value}" for index, value in valores_seguinte_integridade.items()])

print(formatted_output)

LIBERDADE>DIREITOS INDIVIDUAIS>AUTONOMIA DE VONTADE 4057
LIBERDADE>SEXUAL>FISICA>ESTUPRO 3549
LIBERDADE>SEXUAL>PSIQUICA>ASSEDIO SEXUAL 2415
LIBERDADE>DIREITOS INDIVIDUAIS>CARCERE PRIVADO 2302
LIBERDADE>DIREITOS INDIVIDUAIS>LIBERDADE DE IR VIR PERMANECER 1417
LIBERDADE>SEXUAL>FISICA>ABUSO / IMPORTUNACAO SEXUAL FISICA 1398
LIBERDADE>SEXUAL>PSIQUICA>ABUSO / IMPORTUNACAO SEXUAL PSIQUICA 1274
LIBERDADE>SEXUAL>PSIQUICA>ABUSO SEXUAL PSIQUICO 1255
LIBERDADE>SEXUAL>FISICA>ABUSO SEXUAL FISICO 1237
LIBERDADE>DIREITOS INDIVIDUAIS>STALKING 983
LIBERDADE>SEXUAL>IMPORTUNACAO SEXUAL 598
LIBERDADE>SEXUAL>FISICA>EXPLORACAO SEXUAL 354
LIBERDADE>DE RELIGIAO OU CRENCA>DE CRENCA 329
LIBERDADE>LABORAL>EXPLORACAO DO TRABALHO 239
LIBERDADE>EXPRESSAO>LIBERDADE DE CONSCIENCIA E DE PENSAMENTO 186
LIBERDADE>DIREITOS INDIVIDUAIS>CONDICAO ANALOGA A DE ESCRAVO>SUBMETER TRABALHADOR A JORNADA EXAUSTIVA 164
LIBERDADE>SEXUAL>ESTUPRO DE VULNERAVEL 119
LIBERDADE>DE RELIGIAO OU CRENCA>DE CULTO 113
LIBERDADE>DIREITOS INDIVID

In [8]:
def get_full_text(text):
    if isinstance(text, str) and (text.startswith("LIBERDADE") or text.startswith("INTEGRIDADE")):
        return text  # Retorna a string completa
    return None

# Aplicando a função para capturar as strings inteiras que começam com "LIBERDADE" ou "INTEGRIDADE"
df['liberdade_integridade_completa'] = df['Violações'].apply(get_full_text)

# Exibindo os valores que começam com "LIBERDADE" ou "INTEGRIDADE"
valores_liberdade_integridade = df['liberdade_integridade_completa'].dropna().value_counts()
print(valores_liberdade_integridade)


liberdade_integridade_completa
INTEGRIDADE>PSIQUICA>AMEACA OU COACAO                                                                             55538
INTEGRIDADE>PSIQUICA>TORTURA PSIQUICA                                                                             52095
INTEGRIDADE>PSIQUICA>CONSTRANGIMENTO                                                                              40860
INTEGRIDADE>FISICA>AGRESSAO OU VIAS DE FATO                                                                       37343
INTEGRIDADE>PSIQUICA>INJURIA                                                                                      27822
                                                                                                                  ...  
INTEGRIDADE>FISICA>AGRESSAO OU VIAS DE FATO,INTEGRIDADE>FISICA>EXPOSICAO DE RISCO A SAUDE,INTEGRIDADE>PSIQUICA        1
INTEGRIDADE>FISICA>EXPOSICAO DE RISCO A SAUDE,INTEGRIDADE>PSIQUICA>AMEACA OU COACAO,INTEGRIDADE>PSIQUICA              1
INTEGRIDA

In [27]:
cat_violacoes = df['violação_agrupada'].dropna().value_counts()
print(cat_violacoes)

violação_agrupada
Não categorizado                     839815
Violência Psicológica                107723
Violência Física                      83625
Violência Sexual                      58672
Violação da Liberdade e Autonomia     25752
Violência Patrimonial                 11793
Negligência                            1298
Name: count, dtype: int64


In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1128678 entries, 0 to 1128677
Data columns (total 81 columns):
 #   Column                                            Non-Null Count    Dtype  
---  ------                                            --------------    -----  
 0   Data de cadastro                                  1128678 non-null  object 
 1   Canal de atendimento                              592610 non-null   object 
 2   Denúncia emergencial                              461710 non-null   object 
 3   Denunciante                                       946341 non-null   object 
 4   Cenário da violação                               590210 non-null   object 
 5   País                                              403786 non-null   object 
 6   UF                                                889787 non-null   object 
 7   Município                                         871949 non-null   object 
 8   Frequência                                        784879 non-null   obje

In [44]:
cat_violacoes = df['violação_agrupada'].dropna().value_counts()
print(cat_violacoes)

violação_agrupada
Não categorizado                     1022366
Violência Sexual                       50430
Violência Física                       37343
Violação da Liberdade e Autonomia      18240
Violência Patrimonial                    299
Name: count, dtype: int64


In [8]:
def agrupar_violacoes(df):
    # Dicionário de categorias de violência
    violencias = {
        "violencia_fisica": [
            "LESAO CORPORAL LEVE",
            "LESAO CORPORAL GRAVE",
            "LESAO CORPORAL GRAVISSIMA",
            "AGRESSAO OU VIAS DE FATO",
            "MAUS TRATOS",
            "VIOLENCIA OBSTETRICA",
            "EXPOSICAO DE RISCO A SAUDE",
            "HOMICIDIO",
            "DIREITOS SOCIAIS",
            "AGRESSAO OU VIAS DE FATO, INTEGRIDADE>FISICA>EXPOSICAO DE RISCO A SAUDE",
            "AGRESSAO OU VIAS DE FATO, INTEGRIDADE>FISICA>LESAO CORPORAL",
        ],
        "violencia_psicologica": [
            "AMEACA",
            "VIOLENCIA_PSICOLOGICA",
            "INJURIA",
            "DIFAMACAO",
            "CALUNIA",
            "TORTURA PSIQUICA",
            "CONSTRANGIMENTO",
            "INSUBSISTENCIA AFETIVA",
            "ASSEDIO MORAL",
            "BULLYING",
            "EXPOSICAO",
            "ALIENACAO PARENTAL",
        ],
        "violencia_sexual": [
            "ESTUPRO",
            "ASSSEDIO SEXUAL (TRABALHO)",
            "VIOLENCIA SEXUAL",
            "EXPLORACAO SEXUAL",
            "ABUSO / IMPORTUNACAO SEXUAL FISICA",
            "ABUSO / IMPORTUNACAO SEXUAL PSIQUICA",
            "ESTUPRO DE VULNERAVEL",
            "ESTUPRO VIRTUAL",
            "VIOLENCIA SEXUAL MEDIANTE FRAUDE",
        ],
        "violencia_patrimonial": [
            "VIOLENCIA PATRIMONIAL",
            "NEGLIGENCIA",
            "INSUBSISTENCIA MATERIAL",
            "PATRIMONIAL>INDIVIDUAL",
            "PATRIMONIAL>COLETIVO",
            "PATRIMONIAL>CULTURAL",
        ],
        "outro": [
            "OUTROS",
            "OUTRO"
        ],
        "violencia_institucional": [
            "VIOLENCIA INSTITUCIONAL",
            "VIOLENCIA POLICIAL",
            "VIOLENCIA VIRTUAL",
            "TRAFICO DE MULHERES",
            "TRABALHO ESCRAVO",
            "SEQUESTRO",
            "CONDIÇÃO ANALOGA A DE ESCRAVO",
            "FALTA DE ACESSIBILIDADE",
        ],
    }

    # Função para determinar a categoria de violência
    def determinar_categoria(violacao):
        # Verifica se a violação é uma string
        if isinstance(violacao, str):
            for categoria, palavras in violencias.items():
                if any(palavra in violacao for palavra in palavras):
                    return categoria
        return None

    # Aplicar a função à coluna "Violações" e criar a nova coluna "violacao_agrupada"
    df['violacao_agrupada'] = df['Violações'].apply(determinar_categoria)

    return df

In [9]:
# Aplicar a função ao DataFrame
df_atualizado = agrupar_violacoes(df)

# Exibir a contagem das categorias
print(df_atualizado["violacao_agrupada"].value_counts())

violacao_agrupada
violencia_psicologica      439050
violencia_fisica           348781
violencia_patrimonial       64683
violencia_sexual            38505
outro                       14434
violencia_institucional      5247
Name: count, dtype: int64


In [4]:
#linhas_nao_categorizadas = df[df['violação_agrupada'] == "Não categorizado"]

#print(linhas_nao_categorizadas[['Violações', 'violação_agrupada']].sample(n=10))

                                                Violações violação_agrupada
628854  VIOLENCIA DOMESTICA E FAMILIAR.VIOLENCIA PSICO...  Não categorizado
51493                  INTEGRIDADE>PSIQUICA>ASSEDIO MORAL  Não categorizado
637822               FEMINICIDIO.TENTATIVA DE FEMINICIDIO  Não categorizado
502821   DANO EMOCIONAL/DIMINUICAO DA AUTO-ESTIMA; AMEACA  Não categorizado
277997                                LESAO CORPORAL LEVE  Não categorizado
589946                                LESAO CORPORAL LEVE  Não categorizado
9951                              INTEGRIDADE>NEGLIGENCIA  Não categorizado
444486  AMEACA; TENTATIVA DE HOMICIDIO; LESAO CORPORAL...  Não categorizado
260861  INJURIA; CALUNIA; AMEACA; LESAO CORPORAL LEVE;...  Não categorizado
436587   AMEACA; DANO EMOCIONAL/DIMINUICAO DA AUTO-ESTIMA  Não categorizado


In [23]:
import pandas as pd

# Dicionário de categorias de violência
violencias2 = {
    "violencia_fisica": [
        "LESAO CORPORAL LEVE",
        "LESAO CORPORAL GRAVE",
        "LESAO CORPORAL GRAVISSIMA",
        "AGRESSAO OU VIAS DE FATO",
        "MAUS TRATOS",
        "VIOLENCIA OBSTETRICA",
        "EXPOSICAO DE RISCO A SAUDE",
        "HOMICIDIO",
        "VIOLENCIA POLICIAL",
        "DIREITOS SOCIAIS",
        "OUTROS",
        "OUTRO",
        "AGRESSAO OU VIAS DE FATO, INTEGRIDADE>FISICA>EXPOSICAO DE RISCO A SAUDE",
        "AGRESSAO OU VIAS DE FATO, INTEGRIDADE>FISICA>LESAO CORPORAL",
    ],
    "violencia_sexual": [
        "ESTUPRO",
        "ASSSEDIO SEXUAL (TRABALHO)",
        "VIOLENCIA SEXUAL",
        "EXPLORACAO SEXUAL",
        "ABUSO / IMPORTUNACAO SEXUAL FISICA",
        "ABUSO / IMPORTUNACAO SEXUAL PSIQUICA",
        "ESTUPRO DE VULNERAVEL",
        "ESTUPRO VIRTUAL",
        "VIOLENCIA SEXUAL MEDIANTE FRAUDE",
    ],
    # Outras categorias...
}

# Função para contar palavras da categoria 'violencia_sexual' em 'violencia_fisica'
def contar_violencia_sexual_em_fisica(df):
    # Filtrar apenas as linhas que têm 'violencia_fisica'
    df_fisica = df[df['violacao_agrupada'] == 'violencia_fisica']
    
    # Contar as ocorrências das palavras da categoria 'violencia_sexual'
    contagem = {}
    for palavra in violencias2['violencia_sexual']:
        contagem[palavra] = df_fisica['Violações'].str.contains(palavra, na=False).sum()
    
    return contagem

# Aplicar a função ao DataFrame
contagem_violencia_sexual = contar_violencia_sexual_em_fisica(df_atualizado)

# Exibir a contagem
print(contagem_violencia_sexual)


/tmp/ipykernel_20339/1698703185.py:43: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  contagem[palavra] = df_fisica['Violações'].str.contains(palavra, na=False).sum()


{'ESTUPRO': np.int64(23025), 'ASSSEDIO SEXUAL (TRABALHO)': np.int64(0), 'VIOLENCIA SEXUAL': np.int64(0), 'EXPLORACAO SEXUAL': np.int64(51), 'ABUSO / IMPORTUNACAO SEXUAL FISICA': np.int64(0), 'ABUSO / IMPORTUNACAO SEXUAL PSIQUICA': np.int64(0), 'ESTUPRO DE VULNERAVEL': np.int64(0), 'ESTUPRO VIRTUAL': np.int64(0), 'VIOLENCIA SEXUAL MEDIANTE FRAUDE': np.int64(0)}


In [10]:
# Dicionário de categorias de violência
violencias2 = {
    "violencia_fisica": [
        "LESAO CORPORAL LEVE",
        "LESAO CORPORAL GRAVE",
        "LESAO CORPORAL GRAVISSIMA",
        "AGRESSAO OU VIAS DE FATO",
        "MAUS TRATOS",
        "VIOLENCIA OBSTETRICA",
        "EXPOSICAO DE RISCO A SAUDE",
        "HOMICIDIO",
        "DIREITOS SOCIAIS",
        "AGRESSAO OU VIAS DE FATO, INTEGRIDADE>FISICA>EXPOSICAO DE RISCO A SAUDE",
        "AGRESSAO OU VIAS DE FATO, INTEGRIDADE>FISICA>LESAO CORPORAL",
    ],
    "violencia_psicologica": [
        "AMEACA",
        "VIOLENCIA_PSICOLOGICA",
        "INJURIA",
        "DIFAMACAO",
        "CALUNIA",
        "TORTURA PSIQUICA",
        "CONSTRANGIMENTO",
        "INSUBSISTENCIA AFETIVA",
        "ASSEDIO MORAL",
        "BULLYING",
        "EXPOSICAO",
        "ALIENACAO PARENTAL",
    ],
    "violencia_sexual": [
        "ESTUPRO",
        "ASSSEDIO SEXUAL (TRABALHO)",
        "VIOLENCIA SEXUAL",
        "EXPLORACAO SEXUAL",
        "ABUSO / IMPORTUNACAO SEXUAL FISICA",
        "ABUSO / IMPORTUNACAO SEXUAL PSIQUICA",
        "ESTUPRO DE VULNERAVEL",
        "ESTUPRO VIRTUAL",
        "VIOLENCIA SEXUAL MEDIANTE FRAUDE",
    ],
    "violencia_patrimonial": [
        "VIOLENCIA PATRIMONIAL",
        "NEGLIGENCIA",
        "INSUBSISTENCIA MATERIAL",
        "PATRIMONIAL>INDIVIDUAL",
        "PATRIMONIAL>COLETIVO",
        "PATRIMONIAL>CULTURAL",
    ],
    "violencia_institucional": [
        "VIOLENCIA INSTITUCIONAL",
        "VIOLENCIA POLICIAL",
        "VIOLENCIA VIRTUAL",
        "TRAFICO DE MULHERES",
        "TRABALHO ESCRAVO",
        "SEQUESTRO",
        "CONDIÇÃO ANALOGA A DE ESCRAVO",
        "FALTA DE ACESSIBILIDADE",
    ],
}

In [6]:
def contar_violencia_sexual_por_categoria(df):
    # Criar um dicionário para armazenar os resultados
    contagem_total = {categoria: {} for categoria in violencias2.keys() if categoria != 'violencia_sexual'}
    
    # Filtrar linhas que não são da categoria 'violencia_sexual'
    df_outros = df[~df['violacao_agrupada'].isin(['violencia_sexual'])]
    
    # Contar as ocorrências das palavras da categoria 'violencia_sexual' para cada categoria
    for categoria in contagem_total.keys():
        # Filtrar o DataFrame para a categoria atual
        df_categoria = df_outros[df_outros['violacao_agrupada'] == categoria]
        
        for palavra in violencias2['violencia_sexual']:
            # Contar a ocorrência da palavra na categoria atual
            contagem = df_categoria['Violações'].str.contains(palavra, na=False).sum()
            contagem_total[categoria][palavra] = contagem
            
            # Debug: Exibir contagem para cada palavra e categoria
            print(f"Categoria: {categoria}, Palavra: {palavra}, Contagem: {contagem}")
    
    return contagem_total

# Aplicar a função ao DataFrame
contagem_violencia_sexual_por_categoria_resultado = contar_violencia_sexual_por_categoria(df_atualizado)

# Exibir a contagem
print(contagem_violencia_sexual_por_categoria_resultado)


Categoria: violencia_fisica, Palavra: ESTUPRO, Contagem: 23025
Categoria: violencia_fisica, Palavra: ASSSEDIO SEXUAL (TRABALHO), Contagem: 0


/tmp/ipykernel_39222/631927839.py:15: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  contagem = df_categoria['Violações'].str.contains(palavra, na=False).sum()


Categoria: violencia_fisica, Palavra: VIOLENCIA SEXUAL, Contagem: 0
Categoria: violencia_fisica, Palavra: EXPLORACAO SEXUAL, Contagem: 51
Categoria: violencia_fisica, Palavra: ABUSO / IMPORTUNACAO SEXUAL FISICA, Contagem: 0
Categoria: violencia_fisica, Palavra: ABUSO / IMPORTUNACAO SEXUAL PSIQUICA, Contagem: 0
Categoria: violencia_fisica, Palavra: ESTUPRO DE VULNERAVEL, Contagem: 0
Categoria: violencia_fisica, Palavra: ESTUPRO VIRTUAL, Contagem: 0
Categoria: violencia_fisica, Palavra: VIOLENCIA SEXUAL MEDIANTE FRAUDE, Contagem: 0
Categoria: violencia_psicologica, Palavra: ESTUPRO, Contagem: 9283
Categoria: violencia_psicologica, Palavra: ASSSEDIO SEXUAL (TRABALHO), Contagem: 0
Categoria: violencia_psicologica, Palavra: VIOLENCIA SEXUAL, Contagem: 0
Categoria: violencia_psicologica, Palavra: EXPLORACAO SEXUAL, Contagem: 50
Categoria: violencia_psicologica, Palavra: ABUSO / IMPORTUNACAO SEXUAL FISICA, Contagem: 0
Categoria: violencia_psicologica, Palavra: ABUSO / IMPORTUNACAO SEXUAL PSIQ

In [13]:
def contar_violencia_fisica_por_categoria(df):
    # Criar um dicionário para armazenar os resultados
    contagem_total = {categoria: {} for categoria in violencias2.keys() if categoria != 'violencia_fisica'}
    
    # Filtrar linhas que não são da categoria 'violencia_fisica'
    df_outros = df[~df['violacao_agrupada'].isin(['violencia_fisica'])]
    
    # Contar as ocorrências das palavras da categoria 'violencia_fisica' para cada categoria
    for categoria in contagem_total.keys():
        # Filtrar o DataFrame para a categoria atual
        df_categoria = df_outros[df_outros['violacao_agrupada'] == categoria]
        
        for palavra in violencias2['violencia_fisica']:
            # Contar a ocorrência da palavra na categoria atual
            contagem = df_categoria['Violações'].str.contains(palavra, na=False).sum()
            contagem_total[categoria][palavra] = contagem
            
            # Debug: Exibir contagem para cada palavra e categoria
            print(f"Categoria: {categoria}, Palavra: {palavra}, Contagem: {contagem}")
    
    return contagem_total

# Aplicar a função ao DataFrame
contagem_violencia_fisica_por_categoria_resultado = contar_violencia_fisica_por_categoria(df_atualizado)

# Exibir a contagem
print(contagem_violencia_fisica_por_categoria_resultado)


Categoria: violencia_psicologica, Palavra: LESAO CORPORAL LEVE, Contagem: 0
Categoria: violencia_psicologica, Palavra: LESAO CORPORAL GRAVE, Contagem: 0
Categoria: violencia_psicologica, Palavra: LESAO CORPORAL GRAVISSIMA, Contagem: 0
Categoria: violencia_psicologica, Palavra: AGRESSAO OU VIAS DE FATO, Contagem: 0
Categoria: violencia_psicologica, Palavra: MAUS TRATOS, Contagem: 0
Categoria: violencia_psicologica, Palavra: VIOLENCIA OBSTETRICA, Contagem: 0
Categoria: violencia_psicologica, Palavra: EXPOSICAO DE RISCO A SAUDE, Contagem: 0
Categoria: violencia_psicologica, Palavra: HOMICIDIO, Contagem: 0
Categoria: violencia_psicologica, Palavra: DIREITOS SOCIAIS, Contagem: 0
Categoria: violencia_psicologica, Palavra: AGRESSAO OU VIAS DE FATO, INTEGRIDADE>FISICA>EXPOSICAO DE RISCO A SAUDE, Contagem: 0
Categoria: violencia_psicologica, Palavra: AGRESSAO OU VIAS DE FATO, INTEGRIDADE>FISICA>LESAO CORPORAL, Contagem: 0
Categoria: violencia_sexual, Palavra: LESAO CORPORAL LEVE, Contagem: 1787

# dados de Violência Fisica na Sexual

Ocorreram contagens significativas para várias palavras:

- **Lesão Corporal Leve**: 17,870
- **Lesão Corporal Grave**: 3,627
- **Lesão Corporal Gravíssima**: 853
- **Maus Tratos**: 612
- **Exposição de Risco à Saúde**: 367
- **Homicídio**: 2,514


In [14]:
def contar_violencia_patrimonial_por_categoria(df):
    # Criar um dicionário para armazenar os resultados
    contagem_total = {categoria: {} for categoria in violencias2.keys() if categoria != 'violencia_patrimonial'}
    
    # Filtrar linhas que não são da categoria 'violencia_patrimonial'
    df_outros = df[~df['violacao_agrupada'].isin(['violencia_patrimonial'])]
    
    # Contar as ocorrências das palavras da categoria 'violencia_patrimonial' para cada categoria
    for categoria in contagem_total.keys():
        # Filtrar o DataFrame para a categoria atual
        df_categoria = df_outros[df_outros['violacao_agrupada'] == categoria]
        
        for palavra in violencias2['violencia_patrimonial']:
            # Contar a ocorrência da palavra na categoria atual
            contagem = df_categoria['Violações'].str.contains(palavra, na=False).sum()
            contagem_total[categoria][palavra] = contagem
            
            # Debug: Exibir contagem para cada palavra e categoria
            print(f"Categoria: {categoria}, Palavra: {palavra}, Contagem: {contagem}")
    
    return contagem_total

# Aplicar a função ao DataFrame para violência patrimonial
contagem_violencia_patrimonial_por_categoria_resultado = contar_violencia_patrimonial_por_categoria(df_atualizado)

# Exibir a contagem
print(contagem_violencia_patrimonial_por_categoria_resultado)


Categoria: violencia_fisica, Palavra: VIOLENCIA PATRIMONIAL, Contagem: 73560
Categoria: violencia_fisica, Palavra: NEGLIGENCIA, Contagem: 296
Categoria: violencia_fisica, Palavra: INSUBSISTENCIA MATERIAL, Contagem: 1825
Categoria: violencia_fisica, Palavra: PATRIMONIAL>INDIVIDUAL, Contagem: 0
Categoria: violencia_fisica, Palavra: PATRIMONIAL>COLETIVO, Contagem: 0
Categoria: violencia_fisica, Palavra: PATRIMONIAL>CULTURAL, Contagem: 0
Categoria: violencia_psicologica, Palavra: VIOLENCIA PATRIMONIAL, Contagem: 35907
Categoria: violencia_psicologica, Palavra: NEGLIGENCIA, Contagem: 1
Categoria: violencia_psicologica, Palavra: INSUBSISTENCIA MATERIAL, Contagem: 777
Categoria: violencia_psicologica, Palavra: PATRIMONIAL>INDIVIDUAL, Contagem: 0
Categoria: violencia_psicologica, Palavra: PATRIMONIAL>COLETIVO, Contagem: 0
Categoria: violencia_psicologica, Palavra: PATRIMONIAL>CULTURAL, Contagem: 0
Categoria: violencia_sexual, Palavra: VIOLENCIA PATRIMONIAL, Contagem: 20509
Categoria: violencia

# Violência Patrimonial por Categoria

#### Violência Física
- **Violência Patrimonial:** 73,560
- **Negligência:** 296
- **Insubistência Material:** 1,825
- **Patrimonial > Individual:** 0
- **Patrimonial > Coletivo:** 0
- **Patrimonial > Cultural:** 0

#### Violência Psicológica
- **Violência Patrimonial:** 35,907
- **Negligência:** 1
- **Insubistência Material:** 777
- **Patrimonial > Individual:** 0
- **Patrimonial > Coletivo:** 0
- **Patrimonial > Cultural:** 0

#### Violência Sexual
- **Violência Patrimonial:** 20,509
- **Negligência:** 0
- **Insubistência Material:** 102
- **Patrimonial > Individual:** 0
- **Patrimonial > Coletivo:** 0
- **Patrimonial > Cultural:** 0


In [18]:
# Filtrando palavras de interesse
palavras_patrimoniais = ['VIOLENCIA PATRIMONIAL', 'NEGLIGENCIA', 'INSUBSISTENCIA MATERIAL']

# Preencher valores NaN com uma string vazia para evitar erro
df['Violações'] = df['Violações'].fillna('')

# Filtrar o DataFrame para incluir apenas as linhas com palavras patrimoniais
filtered_df = df[df['Violações'].str.contains('|'.join(palavras_patrimoniais))]

# Obter as categorias de violação
outros_tipos = ['violencia_fisica', 'violencia_psicologica', 'violencia_sexual', 'violencia_institucional']

# Armazenando resultados
resultados = []

for index, row in filtered_df.iterrows():
    # Verificar se a linha filtrada tem uma categoria diferente de 'violencia_patrimonial'
    for tipo in outros_tipos:
        if row['violacao_agrupada'] == tipo:
            resultados.append(row['Violações'])

# Contabilizando as palavras
contagem = pd.Series(resultados).value_counts()

# Mostrando os resultados
print(contagem)


VIOLENCIA DOMESTICA E FAMILIAR.VIOLENCIA FISICA;VIOLENCIA DOMESTICA E FAMILIAR.VIOLENCIA MORAL;VIOLENCIA DOMESTICA E FAMILIAR.VIOLENCIA PATRIMONIAL;VIOLENCIA DOMESTICA E FAMILIAR.VIOLENCIA PSICOLOGICA;VIOLENCIA DOMESTICA E FAMILIAR.VIOLENCIA SEXUAL    2955
AMEACA; VIOLENCIA PATRIMONIAL                                                                                                                                                                                                                                854
VIOLENCIA PATRIMONIAL; AMEACA                                                                                                                                                                                                                                796
LESAO CORPORAL LEVE; VIOLENCIA PATRIMONIAL                                                                                                                                                                                           

In [16]:
def contar_violencia_psicologica_por_categoria(df):
    # Criar um dicionário para armazenar os resultados
    contagem_total = {categoria: {} for categoria in violencias2.keys() if categoria != 'violencia_psicologica'}
    
    # Filtrar linhas que não são da categoria 'violencia_psicologica'
    df_outros = df[~df['violacao_agrupada'].isin(['violencia_psicologica'])]
    
    # Contar as ocorrências das palavras da categoria 'violencia_psicologica' para cada categoria
    for categoria in contagem_total.keys():
        # Filtrar o DataFrame para a categoria atual
        df_categoria = df_outros[df_outros['violacao_agrupada'] == categoria]
        
        for palavra in violencias2['violencia_psicologica']:
            # Contar a ocorrência da palavra na categoria atual
            contagem = df_categoria['Violações'].str.contains(palavra, na=False).sum()
            contagem_total[categoria][palavra] = contagem
            
            # Debug: Exibir contagem para cada palavra e categoria
            print(f"Categoria: {categoria}, Palavra: {palavra}, Contagem: {contagem}")
    
    return contagem_total

# Aplicar a função ao DataFrame para violência psicológica
contagem_violencia_psicologica_por_categoria_resultado = contar_violencia_psicologica_por_categoria(df_atualizado)

# Exibir a contagem
print(contagem_violencia_psicologica_por_categoria_resultado)


Categoria: violencia_fisica, Palavra: AMEACA, Contagem: 152935
Categoria: violencia_fisica, Palavra: VIOLENCIA_PSICOLOGICA, Contagem: 17309
Categoria: violencia_fisica, Palavra: INJURIA, Contagem: 111012
Categoria: violencia_fisica, Palavra: DIFAMACAO, Contagem: 96584
Categoria: violencia_fisica, Palavra: CALUNIA, Contagem: 80562
Categoria: violencia_fisica, Palavra: TORTURA PSIQUICA, Contagem: 9500
Categoria: violencia_fisica, Palavra: CONSTRANGIMENTO, Contagem: 14167
Categoria: violencia_fisica, Palavra: INSUBSISTENCIA AFETIVA, Contagem: 4847
Categoria: violencia_fisica, Palavra: ASSEDIO MORAL, Contagem: 6242
Categoria: violencia_fisica, Palavra: BULLYING, Contagem: 0
Categoria: violencia_fisica, Palavra: EXPOSICAO, Contagem: 32215
Categoria: violencia_fisica, Palavra: ALIENACAO PARENTAL, Contagem: 0
Categoria: violencia_sexual, Palavra: AMEACA, Contagem: 25780
Categoria: violencia_sexual, Palavra: VIOLENCIA_PSICOLOGICA, Contagem: 1931
Categoria: violencia_sexual, Palavra: INJURIA, C

# Violencia psicologica em outras categorias

### Violência Física
- **Ameaça:** 152,935
- **Violência Psicológica:** 17,309
- **Injúria:** 111,012
- **Difamação:** 96,584
- **Calúnia:** 80,562
- **Tortura Psíquica:** 9,500
- **Constrangimento:** 14,167
- **Insubistência Afetiva:** 4,847
- **Assédio Moral:** 6,242
- **Bullying:** 0
- **Exposição:** 32,215
- **Alienação Parental:** 0

### Violência Sexual
- **Ameaça:** 25,780
- **Violência Psicológica:** 1,931
- **Injúria:** 18,616
- **Difamação:** 16,731
- **Calúnia:** 13,923
- **Tortura Psíquica:** 1,019
- **Constrangimento:** 1,287
- **Insubistência Afetiva:** 266
- **Assédio Moral:** 674
- **Bullying:** 0
- **Exposição:** 978
- **Alienação Parental:** 0


In [19]:
# Filtrando palavras de interesse para violência psicológica
palavras_psicologicas = [
    'AMEACA',
    'VIOLENCIA_PSICOLOGICA',
    'INJURIA',
    'DIFAMACAO',
    'CALUNIA',
    'TORTURA PSIQUICA',
    'CONSTRANGIMENTO',
    'INSUBSISTENCIA AFETIVA',
    'ASSEDIO MORAL',
    'BULLYING',
    'EXPOSICAO',
    'ALIENACAO PARENTAL'
]

# Preencher valores NaN com uma string vazia para evitar erro
df['Violações'] = df['Violações'].fillna('')

# Filtrar o DataFrame para incluir apenas as linhas com palavras psicológicas
filtered_df_psicologica = df[df['Violações'].str.contains('|'.join(palavras_psicologicas))]

# Obter as categorias de violação
outros_tipos = ['violencia_fisica', 'violencia_sexual', 'violencia_institucional']

# Armazenando resultados
resultados_psicologicos = []

for index, row in filtered_df_psicologica.iterrows():
    # Verificar se a linha filtrada tem uma categoria diferente de 'violencia_psicologica'
    for tipo in outros_tipos:
        if row['violacao_agrupada'] == tipo:
            resultados_psicologicos.append(row['Violações'])

# Contabilizando as palavras
contagem_psicologica = pd.Series(resultados_psicologicos).value_counts()

# Mostrando os resultados
print("Contagem de Violações Psicológicas em Outras Categorias:")
print(contagem_psicologica)


Contagem de Violações Psicológicas em Outras Categorias:
INTEGRIDADE>FISICA>EXPOSICAO DE RISCO A SAUDE                                                                                                      18730
LESAO CORPORAL LEVE; AMEACA                                                                                                                         3873
AMEACA; LESAO CORPORAL LEVE                                                                                                                         3837
DANO EMOCIONAL/DIMINUICAO DA AUTO-ESTIMA; AMEACA; LESAO CORPORAL LEVE                                                                               1867
LESAO CORPORAL LEVE; AMEACA; DANO EMOCIONAL/DIMINUICAO DA AUTO-ESTIMA                                                                               1844
                                                                                                                                                   ...  
CALUNIA; AMEACA; LESAO CO

In [11]:
def contar_violencia_institucional_por_categoria(df):
    # Criar um dicionário para armazenar os resultados
    contagem_total = {categoria: {} for categoria in violencias2.keys() if categoria != 'violencia_institucional'}
    
    # Filtrar linhas que não são da categoria 'violencia_institucional'
    df_outros = df[~df['violacao_agrupada'].isin(['violencia_institucional'])]
    
    # Contar as ocorrências das palavras da categoria 'violencia_institucional' para cada categoria
    for categoria in contagem_total.keys():
        # Filtrar o DataFrame para a categoria atual
        df_categoria = df_outros[df_outros['violacao_agrupada'] == categoria]
        
        for palavra in violencias2['violencia_institucional']:
            # Contar a ocorrência da palavra na categoria atual
            contagem = df_categoria['Violações'].str.contains(palavra, na=False).sum()
            contagem_total[categoria][palavra] = contagem
            
            # Debug: Exibir contagem para cada palavra e categoria
            print(f"Categoria: {categoria}, Palavra: {palavra}, Contagem: {contagem}")
    
    return contagem_total

# Aplicar a função ao DataFrame
contagem_violencia_institucional_por_categoria_resultado = contar_violencia_institucional_por_categoria(df_atualizado)

# Exibir a contagem
print(contagem_violencia_institucional_por_categoria_resultado)


Categoria: violencia_fisica, Palavra: VIOLENCIA INSTITUCIONAL, Contagem: 0
Categoria: violencia_fisica, Palavra: VIOLENCIA POLICIAL, Contagem: 0
Categoria: violencia_fisica, Palavra: VIOLENCIA VIRTUAL, Contagem: 0
Categoria: violencia_fisica, Palavra: TRAFICO DE MULHERES, Contagem: 0
Categoria: violencia_fisica, Palavra: TRABALHO ESCRAVO, Contagem: 61
Categoria: violencia_fisica, Palavra: SEQUESTRO, Contagem: 1
Categoria: violencia_fisica, Palavra: CONDIÇÃO ANALOGA A DE ESCRAVO, Contagem: 0
Categoria: violencia_fisica, Palavra: FALTA DE ACESSIBILIDADE, Contagem: 1925
Categoria: violencia_psicologica, Palavra: VIOLENCIA INSTITUCIONAL, Contagem: 0
Categoria: violencia_psicologica, Palavra: VIOLENCIA POLICIAL, Contagem: 0
Categoria: violencia_psicologica, Palavra: VIOLENCIA VIRTUAL, Contagem: 0
Categoria: violencia_psicologica, Palavra: TRAFICO DE MULHERES, Contagem: 0
Categoria: violencia_psicologica, Palavra: TRABALHO ESCRAVO, Contagem: 24
Categoria: violencia_psicologica, Palavra: SEQU

In [20]:
df_atualizado.loc[df['Violações'].str.contains('ESTUPRO', na=False) & (df['violacao_agrupada'] == 'violencia_fisica'), 'violacao_agrupada'] = 'violencia_sexual'
df_atualizado.loc[df['Violações'].str.contains('Exploração Sexual', na=False) & (df['violacao_agrupada'] == 'violencia_fisica'), 'violacao_agrupada'] = 'violencia_sexual'

df_atualizado.loc[df['Violações'].str.contains('ESTUPRO', na=False) & (df['violacao_agrupada'] == 'violencia_psicologica'), 'violacao_agrupada'] = 'violencia_sexual'
df_atualizado.loc[df['Violações'].str.contains('Exploração Sexual', na=False) & (df['violacao_agrupada'] == 'violencia_psicologica'), 'violacao_agrupada'] = 'violencia_sexual'



# Atualização para 'violencia_fisica'
df_atualizado.loc[df['Violações'].str.contains('FALTA DE ACESSIBILIDADE', na=False) & (df['violacao_agrupada'] == 'violencia_institucional'), 'violacao_agrupada'] = 'violencia_fisica'

# Atualização para 'violencia_psicologica'
df_atualizado.loc[df['Violações'].str.contains('FALTA DE ACESSIBILIDADE', na=False) & (df['violacao_agrupada'] == 'violencia_institucional'), 'violacao_agrupada'] = 'violencia_psicologica'

# Atualização para 'violencia_sexual'
df_atualizado.loc[df['Violações'].str.contains('TRAFICO DE MULHERES', na=False) & (df['violacao_agrupada'] == 'violencia_institucional'), 'violacao_agrupada'] = 'violencia_sexual'

# Atualização para 'violencia_patrimonial'
df_atualizado.loc[df['Violações'].str.contains('FALTA DE ACESSIBILIDADE', na=False) & (df['violacao_agrupada'] == 'violencia_institucional'), 'violacao_agrupada'] = 'violencia_patrimonial'


# Exibir a contagem das categorias
print(df_atualizado["violacao_agrupada"].value_counts())

violacao_agrupada
violencia_psicologica      429437
violencia_fisica           326543
violencia_sexual            70417
violencia_patrimonial       64683
outro                       14434
violencia_institucional      5186
Name: count, dtype: int64
